examples/flutter_gallery/lib/demo/material/bottom_app_bar_demo.dart

```diff
@@ -0,0 +1,571 @@
+  'The BottomAppBar adapts its shape appropriately.';
+
+// Whether the Bottom App Bar's menu should keep icons away from the center or from the end of the screen.
+//
+// When the Floating Action Button is positioned at the end of the screen,
+// it would cover icons at the end of the screen, so the END_FAB mode tells 
+// the MyBottomAppBar to place icons away from the end.
+//
+// Similar logic applies to the CENTER_FAB mode.
+enum _BabMode {
+  END_FAB,
+  CENTER_FAB
+}
+
+// Pairs the Bottom App Bar's menu mode with a Floating Action Button Location.
+class _FabLocationConfiguration {
+  const _FabLocationConfiguration(this.name, this.babMode, this.fabLocation);
+
+  // The name of this configuration.
+  final String name;
+
+  // The _BabMode to place the menu in the bab with.
+  final _BabMode babMode;
+
+  // The location for the Floating Action Button.
+  final FloatingActionButtonLocation fabLocation;
+}
+
+// Map of names to the different shapes of Floating Action Button in this demo.
+class _FabShapeConfiguration {
+  const _FabShapeConfiguration(this.name, this.fab);
+
+  final String name;
+  final Widget fab;
+}
+
+// A bottom app bar with a menu inside it.
+class MyBottomAppBar extends StatelessWidget {
+  const MyBottomAppBar(this.babMode, this.color, this.enableNotch);
+
+  final _BabMode babMode;
+  final Color color;
+  final bool enableNotch;
+
+  final Curve fadeOutCurve = const Interval(0.0, 0.3333);
+  final Curve fadeInCurve = const Interval(0.3333, 1.0);
+
+  @override
+  Widget build(BuildContext context) {
+    final bool showsFirst = babMode == _BabMode.END_FAB;
+    return new BottomAppBar(
+      color: color,
+      hasNotch: enableNotch,
+      child: new Row(
+        children: <Widget> [
+          new IconButton(
+            icon: const Icon(Icons.menu),
+            onPressed: () {
+              showModalBottomSheet<Null>(context: context, builder: (BuildContext context) => const MyDrawer()); },
+          ),
+          new Expanded(
+            child: new AnimatedCrossFade(
+              duration: const Duration(milliseconds: 225),
+              firstChild: babContents(context, _BabMode.END_FAB),
+              firstCurve: showsFirst ? fadeOutCurve  : fadeInCurve,
+              secondChild: babContents(context, _BabMode.CENTER_FAB),
+              secondCurve: showsFirst ? fadeInCurve  : fadeOutCurve,
+              crossFadeState: showsFirst ? CrossFadeState.showFirst : CrossFadeState.showSecond,
+            ),
+          ),
+        ],
+      ),
+    );
+  }
+
+  Widget babContents(BuildContext context, _BabMode babMode) {
+    final List<Widget> rowContents = <Widget> [];
+    if (babMode == _BabMode.CENTER_FAB) {
+      rowContents.add(
+        new Expanded(
+          child: new
+          ConstrainedBox(
+            constraints:
+            const BoxConstraints(maxHeight: 0.0),
+          ),
+        ),
+      );
+    }
+    rowContents.addAll(<Widget> [
```

Comment:
This could be `const <Widget> [` etc

Reply:
The widgets inside are newed because of their empty onPressed callbacks.